checklist/src/main/scala/checklist/Checked.scala

```diff
@@ -0,0 +1,138 @@
+package checklist
+
+import cats.implicits._
+import cats.{Semigroup, Monad, Monoid}
+import cats.data.{NonEmptyList, Ior, Validated}
+
+sealed trait Checked[E, W, R] {
+  def widen[EE >: E, WW >: W, RR >: R]: Checked[EE, WW, RR] = this.asInstanceOf[Checked[EE, WW, RR]]
+  def widenErrors[EE >: E]: Checked[EE, W, R] = this.asInstanceOf[Checked[EE, W, R]]
+  def widenWarnings[WW >: W]: Checked[E, WW, R] = this.asInstanceOf[Checked[E, WW, R]]
+  def widenResult[RR >: R]: Checked[E, W, RR] = this.asInstanceOf[Checked[E, W, RR]]
+
+  def fold[A](err: (NonEmptyList[E], List[W]) => A, warn: (NonEmptyList[W], R) => A, succ: R => A): A =
+    this match {
+      case Errored(e, w) => err(e, w)
+      case Warned(w, r) => warn(w, r)
+      case Succeeded(r) => succ(r)
+    }
+
+  def recover(func: (NonEmptyList[E], List[W]) => R): Checked[E, W, R] = recoverWith((e, w) => Succeeded(func(e, w)))
+
+  def recoverWith(func: (NonEmptyList[E], List[W]) => Checked[E, W, R]): Checked[E, W, R] =
+    this match {
+      case Errored(errs, warns) => func(errs, warns)
+      case other => other
+    }
+
+  def combine(other: Checked[E, W, R])(implicit semigroup: Semigroup[R]): Checked[E, W, R] = combineWith(other)(semigroup.combine)
+
+  def combineWith[S, T](other: Checked[E, W, S])(f: (R, S) => T): Checked[E, W, T] =
```

Comment:
There's probably a better name for this.  I think it's map2

Reply:
Agreed. Yeah I think it's `map2`.